src/Microsoft.AspNetCore.Mvc.Core/ApiExplorer/ApiConventionResult.cs

```diff
@@ -0,0 +1,195 @@
+
+        public IReadOnlyList<IApiResponseMetadataProvider> ResponseMetadataProviders { get; }
+
+        internal static ApiConventionResult GetApiConvention(MethodInfo method, ApiConventionTypeAttribute[] apiConventionAttributes)
+        {
+            foreach (var attribute in apiConventionAttributes)
+            {
+                var conventionMethod = GetConventionMethod(method, attribute.ConventionType);
+                if (conventionMethod != null)
+                {
+                    var metadataProviders = conventionMethod.GetCustomAttributes(inherit: false)
+                        .OfType<IApiResponseMetadataProvider>()
+                        .ToArray();
+
+                    return new ApiConventionResult(metadataProviders);
+                }
+            }
+
+            return null;
+        }
+
+        private static MethodInfo GetConventionMethod(MethodInfo method, Type conventionType)
+        {
+            foreach (var conventionMethod in conventionType.GetMethods())
+            {
+                if (IsMatch(method, conventionMethod))
+                {
+                    return conventionMethod;
+                }
+            }
+
+            return null;
+        }
+
+        internal static bool IsMatch(MethodInfo methodInfo, MethodInfo conventionMethod)
+        {
+            var methodNameMatchBehavior = GetNameMatchBehavior(conventionMethod);
+            if (!IsNameMatch(methodInfo.Name, conventionMethod.Name, methodNameMatchBehavior))
+            {
+                return false;
+            }
+
+            var methodParameters = methodInfo.GetParameters();
+            var conventionMethodParameters = conventionMethod.GetParameters();
+
+            for (var i = 0; i < conventionMethodParameters.Length; i++)
+            {
+                var conventionParameter = conventionMethodParameters[i];
+                if (conventionParameter.IsDefined(typeof(ParamArrayAttribute)))
+                {
+                    return true;
+                }
+
+                if (methodParameters.Length <= i)
+                {
+                    return false;
+                }
+
+                var nameMatchBehavior = GetNameMatchBehavior(conventionParameter);
+                var typeMatchBehavior = GetTypeMatchBehavior(conventionParameter);
+
+                if (!IsTypeMatch(methodParameters[i].ParameterType, conventionParameter.ParameterType, typeMatchBehavior) ||
+                    !IsNameMatch(methodParameters[i].Name, conventionParameter.Name, nameMatchBehavior))
+                {
+                    return false;
+                }
+            }
+
+            // Ensure convention has at least as many parameters as the method. params convention argument are handled
+            // inside the for loop.
+            return methodParameters.Length == conventionMethodParameters.Length;
+
+            ApiConventionNameMatchBehavior GetNameMatchBehavior(ICustomAttributeProvider attributeProvider)
```

Comment:
Should be its own unit tested method. This defines the default behavior of the name match attribute. Should do this for a lot of the sub-methods in this class. There's a lot of underlying logic.